### PR TITLE
Wcohen/efficiency

### PR DESCRIPTION
**Pull request by wcohen**
A pair of patches to use the more efficient sys_enter and sys_exit tracepoints for the "-S" option and update the documentation to reflect the current output of the "-S" option.

### CODE REVIEW

**Comment by yonghong-song** [comment]
@pchaigno  @myllynen @r4f4 You are frequent users for this tool, any comments?

**Comment by myllynen** [comment]
Otherwise I think it makes sense but since tools/syscount.py has the same code for determining syscall numbers I wonder should a new helper be introduced for that purpose?

**Comment by yonghong-song** [comment]
Adding a common helper makes sense. since the file is too big, maybe include it in a separate file `syscall.py` in `src/python/bcc` directory? @wcohen could you try and see whether this will work?

**Comment by wcohen** [comment]
I can factor out the syscall id->name code into a separate helper file.

**Comment by wcohen** [comment]
I put the syscall_name() code into src/python/bcc/syscall.py.  The ucalls.py and  syscount.py scripts are now updated to use that library in the wcohen/efficiency branch.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
@myllynen any comments?

**Comment by myllynen** [comment]
LGTM (apart from the points you already mentioned).

File: src/python/bcc/syscall.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    304: b"open_by_handle_at",
+    305: b"clock_adjtime",
+    306: b"syncfs",
+    307: b"sendmmsg",
+    308: b"setns",
+    309: b"getcpu",
+    310: b"process_vm_readv",
+    311: b"process_vm_writev",
+    312: b"kcmp",
+    313: b"finit_module",
```
[the remark this answers is not in the document]
Comment (wcohen) [reply src/python/bcc/syscall.py]: There is a new commit 3c461b37298a328adc39324a2812f4c232c8d028 on the branch to update the table with these newer syscalls.  There were a few additional new syscalls added and pread, pwrite,  _exit, umount became pread64, pwrite64, exit, and umount2 respectively.

**Comment by yonghong-song** [comment]
@wcohen could you put the new commit (adding additional syscall number->names) to this pull request? Also could you add the license for `syscall.py`?

**Comment by wcohen** [comment]
I have put in a commit adding a license to syscall.py and documenting the syscall number->names changes.  Anything else need to be addressed in the patch series?

**Comment by yonghong-song** [comment]
The patch itself looks good. The test failed on ubuntu 16.04, the following is the call stack and error info.
```
27: Traceback (most recent call last):
27:   File "../../tools/lib/ucalls.py", line 258, in <module>
27:     bpf = BPF(text=program, usdt_contexts=[usdt] if usdt else [])
27:   File "/home/iovisor/jenkins/workspace/bcc-pr/label/ubuntu1604/build/src/python/bcc/__init__.py", line 338, in __init__
27:     self._trace_autoload()
27:   File "/home/iovisor/jenkins/workspace/bcc-pr/label/ubuntu1604/build/src/python/bcc/__init__.py", line 1053, in _trace_autoload
27:     fn = self.load_func(func_name, BPF.TRACEPOINT)
27:   File "/home/iovisor/jenkins/workspace/bcc-pr/label/ubuntu1604/build/src/python/bcc/__init__.py", line 378, in load_func
27:     (func_name, errstr))
27: Exception: Failed to load BPF program tracepoint__raw_syscalls__sys_enter: Invalid argument
27: F../bin/sh: 1: ausyscall: not found
27: ./bin/sh: 1: ausyscall: not found
27: 'unknown': I need something more specific.
27: ./bin/sh: 1: ausyscall: not found
27: Traceback (most recent call last):
27:   File "../../tools/lib/uthreads.py", line 131, in <module>
27:     bpf.perf_buffer_poll()
27:   File "/home/iovisor/jenkins/workspace/bcc-pr/label/ubuntu1604/build/src/python/bcc/__init__.py", line 1229, in perf_buffer_poll
27:     lib.perf_reader_poll(len(readers), readers, timeout)
27: KeyboardInterrupt
27: ./bin/sh: 1: ausyscall: not found
27: ./bin/sh: 1: ausyscall: not found
27: .s....
27: ======================================================================
27: FAIL: test_ucalls (__main__.SmokeTests)
27: ----------------------------------------------------------------------
27: Traceback (most recent call last):
27:   File "/home/iovisor/jenkins/workspace/bcc-pr/label/ubuntu1604/tests/python/test_tools_smoke.py", line 348, in test_ucalls
27:     timeout=60, kill_timeout=60)
27:   File "/home/iovisor/jenkins/workspace/bcc-pr/label/ubuntu1604/tests/python/test_tools_smoke.py", line 52, in run_with_int
27:     or (rc == 137 and kill), "rc was %d" % rc)
27: AssertionError: rc was 1
```
Could you take a look?

**Comment by wcohen** [comment]
I am able to reproduce it on x86_64 by removing the /usr/bin/ausyscall binary. Will look at what is causing the problem today.

**Comment by wcohen** [comment]
I think I may have spoken too soon about the ausyscall. I do need to eliminate the stderr "command not found" warning for missing ausyscall. However, the problem jenkins is reporting is releated to the tracepoint.   The ucalls.py needs to be run as root to install the tracepoints on my fedora machine.  How precisely does jenkins run the test?  Were there earlier tests of ucalls.py that passed?

$ ./ucalls.py -l none -S $(pgrep emacs)
/bin/sh: ausyscall: command not found
/virtual/main.c:91:1: warning: declaration of 'struct
      tracepoint__raw_syscalls__sys_enter' will not be visible outside of this
      function [-Wvisibility]
TRACEPOINT_PROBE(raw_syscalls, sys_enter) {
^
/virtual/include/bcc/helpers.h:763:46: note: expanded from macro
      'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__#...
                                             ^
<scratch space>:7:1: note: expanded from here
tracepoint__raw_syscalls__sys_enter
^
/virtual/main.c:93:25: error: incomplete definition of type 'struct
      tracepoint__raw_syscalls__sys_enter'
    u64 *valp, id = args->id, val = 0;
                    ~~~~^
/virtual/main.c:91:1: note: forward declaration of 'struct
      tracepoint__raw_syscalls__sys_enter'
TRACEPOINT_PROBE(raw_syscalls, sys_enter) {
^
/virtual/include/bcc/helpers.h:763:46: note: expanded from macro
      'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__#...
                                             ^
<scratch space>:7:1: note: expanded from here
tracepoint__raw_syscalls__sys_enter
^
1 warning and 1 error generated.
Traceback (most recent call last):
  File "./ucalls.py", line 258, in <module>
    bpf = BPF(text=program, usdt_contexts=[usdt] if usdt else [])
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 319, in __init__
    raise Exception("Failed to compile BPF text")
Exception: Failed to compile BPF text


$ sudo ./ucalls.py -l none -S $(pgrep emacs)
/bin/sh: ausyscall: command not found
Attached kernel tracepoints for syscall tracing.
Tracing calls in process 11600 (language: none)... Ctrl-C to quit.
^C
METHOD                                              # CALLS
getuid                                                    1
lseek                                                     2
write                                                     4
wait4                                                     5
pipe2                                                     5
vfork                                                     5
fstat                                                     5
timerfd_settime                                           7
restart_syscall                                          12
openat                                                   15
read                                                     21
close                                                    25
readlinkat                                               25
tgkill                                                   34
getpid                                                   34
faccessat                                                88
stat                                                    104
writev                                                  188
pselect6                                                205
rt_sigreturn                                            379
rt_sigprocmask                                          878
poll                                                   1343
recvmsg                                                1608
Detaching kernel probes, please wait...

**Comment by yonghong-song** [comment]
I see. kernel 4.4 does not have tracepoint support. It got started at 4.7. You can annotate the test to be executed only if kernel version is >= 4.7.

**Comment by wcohen** [comment]
Where is this test for the test located?  Is there  similar gating for syscount.py needed?  Is is going to fail in pretty much the same way if jenkins run it.

**Comment by wcohen** [comment]
Adjusted the test to only run with linux-4.7 or newer.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Pull request merged**